ui/src/components/CsvSourceMapping.vue

```diff
@@ -65,6 +72,17 @@ export default Vue.extend({
     getColumnMappings () {
       return []
     },
+
+    async deleteSource (source: Source): Promise<void> {
+      const loader = this.$buefy.loading.open({})
+
+      try {
+        await this.$store.dispatch('cubeProjects/deleteSource', source)
+        // TODO: Handle errors
+      } finally {
+        loader.close()
+      }
+    },
```

Comment:
Both are possible, but the loading state is very tied to the current component and not interesting for other components. It also allows to display a "local" loading state (as opposed to full-screen). I don't do that here, but that's an option that you don't really have if your loading state is global.

Reply:
Though I guess for most "delete" actions we could make it generic...

Reply:
Some things that I don't know how to do generically:
- The confirm dialog message
- The success message (which I forgot to do)
- Refetching the collection (or removing the resource from the collection)

Reply:
here's some inspiration

1. set the operation to state as "awaiting confirmation"
2. set callback action name(s) to state
2. hook it up to a `this.$buefy.dialog.confirm` somewhere in the app and show when there is an operation to confirm
3. dispatch a store `invoke` action with parameters `{ operation, actions }`
4. set a loading flag
5. when the operation finishes dispatch the `actions` and push messages to a `messages` state
6. display `messages` state somewhere, like toasts

Reply:
I took some of your suggestions. I didn't go all the way because I kept the confirm dialog in the component. I don't think it's worth it to make it based on global state. This refactor already adds multiple levels of indirection that I'm not sure I like. We'll see.